src/internal/ajax/ajax.ts

```diff
@@ -262,3 +268,189 @@ export const ajax: AjaxCreationMethod = (() => {
+
+      {
+        ///////////////////////////////////////////////////
+        // set up the events before open XHR
+        // https://developer.mozilla.org/en/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest
+        // You need to add the event listeners before calling open() on the request.
+        // Otherwise the progress events will not fire.
+        ///////////////////////////////////////////////////
+
+        const progressSubscriber = config.progressSubscriber;
+
+        xhr.ontimeout = (e: ProgressEvent) => {
+          const timeoutError = new AjaxTimeoutError(xhr, _request);
+          progressSubscriber?.error?.(timeoutError);
+          destination.error(timeoutError);
+        };
+
+        if (progressSubscriber) {
+          xhr.upload.onprogress = (e: ProgressEvent) => {
+            progressSubscriber.next?.(e);
+          };
+        }
+
+        xhr.onerror = (e: ProgressEvent) => {
+          progressSubscriber?.error?.(e);
+          destination.error(new AjaxError('ajax error', xhr, _request));
+        };
+
+        xhr.onload = (e: ProgressEvent) => {
+          // 4xx and 5xx should error (https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)
+          if (xhr.status < 400) {
+            progressSubscriber?.complete?.();
+
+            done = true;
+            let response: AjaxResponse<T>;
+            try {
+              // This can throw in IE, because we end up needing to do a JSON.parse
+              // of the response in some cases to produce object we'd expect from
+              // modern browsers.
+              response = new AjaxResponse(e, xhr, _request);
+            } catch (err) {
+              destination.error(err);
+              return;
+            }
+            destination.next(response);
+            destination.complete();
+          } else {
+            progressSubscriber?.error?.(e);
+            destination.error(new AjaxError('ajax error ' + xhr.status, xhr, _request));
+          }
+        };
+      }
+
+      const { user, method, async } = _request;
+      // open XHR
+      if (user) {
+        xhr.open(method, url, async, user, _request.password);
+      } else {
+        xhr.open(method, url, async);
+      }
+
+      // timeout, responseType and withCredentials can be set once the XHR is open
+      if (async) {
+        xhr.timeout = _request.timeout;
+        xhr.responseType = _request.responseType;
+      }
+
+      if ('withCredentials' in xhr) {
+        xhr.withCredentials = _request.withCredentials;
+      }
+
+      // set headers
+      for (const key in headers) {
+        if (headers.hasOwnProperty(key)) {
+          xhr.setRequestHeader(key, headers[key]);
+        }
+      }
+
+      // finally send the request
+      if (body) {
+        xhr.send(body);
+      } else {
+        xhr.send();
+      }
+    } catch (err) {
+      destination.error(err);
+    }
+
+    return () => {
+      if (!done && xhr) {
```

Comment:
I think `done` could be removed? https://github.com/angular/angular/pull/37367/files#diff-26683cfb34fa97883213282055aef1e7R342

```suggestion
      if (xhr && xhr.readyState !== xhr.DONE) {
```

Reply:
Either that or move `done` outside of the `if` statement, right? Otherwise, I think that it will try to abort requests that have completed with 4xx and 5xx codes.

Reply:
Good point. Thank you.